cpp/include/raft/core/detail/popc.cuh

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/detail/mdspan_util.cuh>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/linalg/coalesced_reduction.cuh>
+
+namespace raft::detail {
+
+/**
+ * @brief Count the number of bits that are set to 1 in a vector.
+ *
+ * @tparam value_t the value type of the vector.
+ * @tparam index_t the index type of vector and scalar.
+ *
+ * @param[in] res raft handle for managing expensive resources
+ * @param[in] values Number of row in the matrix.
+ * @param[in] max_len Maximum number of bits to count.
+ * @param[out] counter Number of bits that are set to 1.
+ */
+template <typename value_t, typename index_t>
+void popc(const raft::resources& res,
+          device_vector_view<value_t, index_t> values,
+          index_t max_len,
```

Comment:
Should this be a `raft::host_scalar_view<index_t>` instead for consistency?

Reply:
It's OK for me, but not so sure, how to judge a primitive type need to be `raft::host_scalar_view<index_t>`？

Reply:
I think that's just the API consistency that we are going for. If it's an array, use vector/matrix_view and if it's a scalar use scalar_view. 

cc @cjnolet for his opinion

Reply:
I think primitive types are okay for values that might never need to originate or be returned on device, as opposed to host.